File: AndroidLab/ch6/src/main/java/com/example/ch6/section3/Test3_1Activity.java
package com.example.ch6.section3;

import android.graphics.Typeface;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.text.Html;
import android.text.SpannableStringBuilder;
import android.text.Spanned;
import android.text.style.ImageSpan;
import android.text.style.RelativeSizeSpan;
import android.text.style.StyleSpan;
import android.text.style.TypefaceSpan;

import androidx.activity.EdgeToEdge;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.res.ResourcesCompat;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;

import com.example.ch6.R;
import com.example.ch6.databinding.ActivityTest31Binding;

public class Test3_1Activity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        EdgeToEdge.enable(this);

        ActivityTest31Binding binding = ActivityTest31Binding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main), (v, insets) -> {
            Insets systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom);
            return insets;
        });

        String data = "복수초 \n img \n 이른봄 설산에서 만나는 복수초는 어쩌구...";

        //문자열 데이터 추가
        //필요한 만큼 각 Span 객체를 추가..
        SpannableStringBuilder builder = new SpannableStringBuilder(data);

        int start = data.indexOf("img");//img 라는 문자열이 있는 최초의 위치..
        if(start > -1){//없으면 -1, 0이상이라면 무조건 있다는 의미..
            int end = start + "img".length();

            //이미지 준비..
            //문자열 데이터 - String
            //이미지 데이터 - Drawable/Bitmap
            Drawable dr = ResourcesCompat.getDrawable(getResources(), R.drawable.img1, null);
            //그리는 좌표 설정..
            dr.setBounds(0, 0, dr.getIntrinsicWidth(), dr.getIntrinsicWidth());
            ImageSpan span = new ImageSpan(dr);
            builder.setSpan(span, start, end, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
        }
        start = data.indexOf("복수초");
        if(start > -1){
            int end = start + "복수초".length();
            StyleSpan styleSpan = new StyleSpan(Typeface.BOLD);
            RelativeSizeSpan sizeSpan = new RelativeSizeSpan(2.0f);
            builder.setSpan(styleSpan, start, end, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
            builder.setSpan(sizeSpan, start, end, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
        }

        binding.text1.setText(builder);


        //html 태그 효과를 적용해서.. 문자열 출력..
        //브라우저로 출력하는 것 아니다.. TextView 에...
        //내부적으로는 Span 클래스 이용한다.. 단지 우리가 안하는 것 뿐이다..
        String html = "<font color='RED'>얼레지</font><br/><img src='img1'/><br/>곰배령...";

        //만약 html 내에 <img> 만 없다면.. 매개변수 하나짜리..
        //<img> 태그가 있다면 그 위치에 이미지가 출력되어야 한다.. 이미지 획득해야 한다.. 그걸 해주지는 않는다.
        //브라우저 아니다..
        //<img> 태그가 있다면 그 위치에 출력될 이미지는 우리 코드로 직접 획득해야 한다..
        //MyImageGetter 가 그 목적의 클래스..
        binding.text2.setText(
                Html.fromHtml(html, Html.FROM_HTML_MODE_LEGACY, new MyImageGetter(), null));
    }

    class MyImageGetter implements Html.ImageGetter{
        //fromHtml() 함수가 분석하는 문자열에 <img> 를 만나면 호출.. 10개면 10번 호출..
        //매개변수로 img 태그의 src 값을 전달해 준다..
        @Override
        public Drawable getDrawable(String source) {
            //서버, 파일, 리소스등 직접 획득해서 리턴.. 리턴 시킨 이미지가 <img> 위치에 출력..
            if(source.equals("img1")){
                Drawable dr = ResourcesCompat.getDrawable(getResources(), R.drawable.img2, null);
                dr.setBounds(0, 0, dr.getIntrinsicWidth(), dr.getIntrinsicHeight());
                return dr;
            }
            //이미지 없다..는 의미..
            return null;
        }
    }
}

File: AndroidLab2/ch1/src/main/java/com/example/ch1/section3_recyclerview/Test3_1Activity.kt
package com.example.ch1.section3_recyclerview

import android.content.Context
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Rect
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.activity.enableEdgeToEdge
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.res.ResourcesCompat
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.StaggeredGridLayoutManager
import com.example.ch1.R
import com.example.ch1.databinding.ActivityTest31Binding
import com.example.ch1.databinding.RecyclerItemBinding

class Test3_1Activity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        enableEdgeToEdge()

        val binding = ActivityTest31Binding.inflate(layoutInflater)
        setContentView(binding.root)

        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main)) { v, insets ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
            insets
        }

        val data = mutableListOf<String>()
        for(i in 1..10){
            data.add("Item $i")
        }

        binding.main.adapter = MyAdapter(data)

        //linear..............세로방향 나열..........
//        binding.main.layoutManager = LinearLayoutManager(this)

        //linear, 가로방향.....
//        val layoutManager = LinearLayoutManager(this)
//        layoutManager.orientation = LinearLayoutManager.HORIZONTAL

        //gridlayout
//        val layoutManager = GridLayoutManager(this, 2)

        //gridlayout, vertical, true - 아래부터 항목 나열..
//        val layoutManager = GridLayoutManager(this, 2,
//            GridLayoutManager.VERTICAL, true)

        //gridlayout, horizontal, false
//        val layoutManager = GridLayoutManager(this, 2,
//            GridLayoutManager.HORIZONTAL, false)

        //staggered.....................
//        data.add("aaaaaaaaaa aaaaaaaaaaaaa, aaaa, aaaaaaaaaaaaaa, aaaaaaaaaaaa, aaaaaaaaaaaaaa, aaaaaaaaaaaa, aaaaaaaaaaaaaaaaa, aaaaaaaaaaaaaaaaa")
//        data.add("bbbbbbbbbb bbbbbbbbbbb")
//        data.add("ccccccccc")
//        data.add("dddddddddddddd ddddddddddddddd dddddddddddddddd ddddddddddddddddd ddddddddddddddddd ddddddddddddddd ddddddddddddddddd ddddddddddddddddddd")
//        val layoutManager = StaggeredGridLayoutManager(2, StaggeredGridLayoutManager.VERTICAL)

        val layoutManager = LinearLayoutManager(this)
        binding.main.layoutManager = layoutManager
        binding.main.addItemDecoration(MyDecoration(this))



    }
}

class MyViewHolder(val binding: RecyclerItemBinding): RecyclerView.ViewHolder(binding.root)

class MyAdapter(val data: MutableList<String>): RecyclerView.Adapter<MyViewHolder>(){
    override fun getItemCount(): Int {
        return data.size
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MyViewHolder {
        return MyViewHolder(RecyclerItemBinding.inflate(LayoutInflater.from(parent.context),
            parent, false))
    }

    override fun onBindViewHolder(holder: MyViewHolder, position: Int) {
        val binding = holder.binding
        binding.itemData.text = data[position]
    }
}

class MyDecoration(val context: Context): RecyclerView.ItemDecoration(){
    //항목이 찍히기 전에 최초에 한번.. 바탕 드로잉이 필요할 때...
    override fun onDraw(c: Canvas, parent: RecyclerView, state: RecyclerView.State) {
        super.onDraw(c, parent, state)
        c.drawBitmap(BitmapFactory.decodeResource(context.resources, R.drawable.stadium),
            0f, 0f, null)
    }

    //항목이 모두 찍힌 후 최후에 한번.. 항목 위에 추가 드로잉이 필요한 경우..
    override fun onDrawOver(c: Canvas, parent: RecyclerView, state: RecyclerView.State) {
        super.onDrawOver(c, parent, state)

        //이미지를 recyclerview 의 center 위치에 그리고 싶다.. 좌표계산..
        //뷰의 크기 계산..
        val width = parent.width
        val height = parent.height

        //드로잉 하고자 하는 이미지 사이즈..
        val dr: Drawable? = ResourcesCompat.getDrawable(context.resources,
            R.drawable.kbo, null)
        val drWidth = dr?.intrinsicWidth
        val drHeight = dr?.intrinsicHeight

        //이미지가 center에 그려지게.. 이미지가 그려질 left, top 좌표값 계산..
        val left = width/2 - drWidth?.div(2) as Int
        val top = height/2 - drHeight?.div(2) as Int

        //이미지 그리기...
        c.drawBitmap(BitmapFactory.decodeResource(context.resources, R.drawable.kbo),
            left.toFloat(), top.toFloat(), null)
    }
    //항목 하나당 한번씩 호출.. 항목에 추가 꾸미기 작업이 필요할 때...
    //outRect : 항목이 출력되는 사각형 정보..
    //view : 항목에 출력되는 뷰 객체..
    override fun getItemOffsets(
        outRect: Rect,
        view: View,
        parent: RecyclerView,
        state: RecyclerView.State
    ) {
        super.getItemOffsets(outRect, view, parent, state)
        //매개변수로 몇번째 항목을 꾸미기 위해서 호출된 것인지, 전달되지 않는다.. 필요하다면 획득 가능..
        //1 을 더하는 것은 알고리즘의 편의성 때문..
        val index = parent.getChildAdapterPosition(view) + 1
        if(index % 3 == 0)
            outRect.set(10, 10, 10, 60)
        else
            outRect.set(10, 10, 10, 0)

        view.setBackgroundColor(Color.LTGRAY)
        ViewCompat.setElevation(view, 20.0f)
    }
}















